src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/FrameConnection.cs

```diff
@@ -275,5 +327,35 @@ private void AssignTimeout(long ticks, TimeoutAction timeoutAction)
             // Add Heartbeat.Interval since this can be called right before the next heartbeat.
             Interlocked.Exchange(ref _timeoutTimestamp, _lastTimestamp + ticks + Heartbeat.Interval.Ticks);
         }
+
+        public void StartTimingReads()
+        {
+            Interlocked.Exchange(ref _readTimingElapsed, 0);
+            Interlocked.Exchange(ref _bytesRead, 0);
+            Interlocked.Exchange(ref _readTimingStartTicks, _lastTimestamp);
+            _timingReads = true;
+        }
+
+        public void StopTimingReads()
+        {
+            _timingReads = false;
+        }
+
+        public void PauseTimingReads()
+        {
+            Log.RequestBodyTimingPause(_context.ConnectionId, _frame.TraceIdentifier, TimeSpan.FromTicks(_readTimingElapsed), TimeSpan.FromTicks(_lastTimestamp - _readTimingStartTicks));
+            _pauseTimingReads = true;
```

Comment:
I'm pretty sure you're going to need to add a lock to FrameConnection. The locks would be per-connection and should always be for quick operations, but I think we could use Monitor.TryEnter in Tick if need be.